### PR TITLE
Implementing Hashing typeclass

### PR DESCRIPTION
**Pull request by axel22**
The same pull request as the previous one (https://github.com/scala/scala/pull/652), to trigger the build bot again.

review by @jsuereth
review by @retronym

### CODE REVIEW

File: src/library/scala/collection/concurrent/TrieMap.scala [hunk 1]
```diff
@@ -894,14 +913,15 @@ object TrieMap extends MutableMapFactory[TrieMap] {
...
 
   def empty[K, V]: TrieMap[K, V] = new TrieMap[K, V]
-
-  @inline final def computeHash[K](k: K): Int = {
-    var hcode = k.hashCode
-    hcode = hcode * 0x9e3775cd
-    hcode = java.lang.Integer.reverseBytes(hcode)
-    hcode * 0x9e3775cd
+  
+  class MangledHashing[K] extends Hashing[K] {
```
Comment (jsuereth) [inline src/library/scala/collection/concurrent/TrieMap.scala:917]: Don't mean to be terribly picky, but should this be a top-level private[concurrent] class?  Avoiding `$outer` in bytecode is ok by me...

**Comment by jsuereth** [comment]
One minor comment, but otherwise LGTM

File: src/library/scala/collection/concurrent/TrieMap.scala (hunk 1 above)
Comment (axel22) [inline src/library/scala/collection/concurrent/TrieMap.scala:917]: No `$outer` pointers as far as I can see - the class is defined in a top-level companion object:

```
Compiled from "TrieMap.scala"
public class scala.collection.concurrent.TrieMap$MangledHashing extends java.lang.Object implements scala.util.hashing.Hashing
  SourceFile: "TrieMap.scala"
  Scala: length = 0x

  Signature: length = 0x2
   00 27 
  InnerClass: 
   public #45= #42 of #44; //MangledHashing=class scala/collection/concurrent/TrieMap$MangledHashing of class scala/collection/concurrent/TrieMap
  minor version: 0
  major version: 49
  Constant pool:
const #1 = Asciz    SourceFile;
const #2 = Asciz    TrieMap.scala;
...
const #39 = Asciz   <K:Ljava/lang/Object;>Ljava/lang/Object;Lscala/util/hashing/Hashing<TK;>;;
const #40 = Asciz   InnerClasses;
const #41 = Asciz   scala/collection/concurrent/TrieMap$MangledHashing;
const #42 = class   #41;    //  scala/collection/concurrent/TrieMap$MangledHashing
const #43 = Asciz   scala/collection/concurrent/TrieMap;
const #44 = class   #43;    //  scala/collection/concurrent/TrieMap
const #45 = Asciz   MangledHashing;
const #46 = Asciz   scala/util/hashing/Hashing;
const #47 = class   #46;    //  scala/util/hashing/Hashing

{
public int hashCode(java.lang.Object);
  LineNumberTable: 
   line 919: 3
   line 920: 11
   line 921: 16

  LocalVariableTable: 
   Start  Length  Slot  Name   Signature
   0      21      0    this       Lscala/collection/concurrent/TrieMap$MangledHashing;
   0      21      1    k       Ljava/lang/Object;
   11      9      2    hcode       I


  Code:
   Stack=2, Locals=3, Args_size=2
   0:   getstatic   #11; //Field scala/runtime/ScalaRunTime$.MODULE$:Lscala/runtime/ScalaRunTime$;
   3:   aload_1
   4:   invokevirtual   #15; //Method scala/runtime/ScalaRunTime$.hash:(Ljava/lang/Object;)I
   7:   ldc #16; //int -1640532531
   9:   imul
   10:  istore_2
   11:  iload_2
   12:  invokestatic    #22; //Method java/lang/Integer.reverseBytes:(I)I
   15:  istore_2
   16:  iload_2
   17:  ldc #16; //int -1640532531
   19:  imul
   20:  ireturn
  LineNumberTable: 
   line 919: 3
   line 920: 11
   line 921: 16

  LocalVariableTable: 
   Start  Length  Slot  Name   Signature
   0      21      0    this       Lscala/collection/concurrent/TrieMap$MangledHashing;
   0      21      1    k       Ljava/lang/Object;
   11      9      2    hcode       I

  Signature: length = 0x2
   00 1E 

public scala.collection.concurrent.TrieMap$MangledHashing();
  LineNumberTable: 
   line 917: 0



  Code:
   Stack=1, Locals=1, Args_size=1
   0:   aload_0
   1:   invokespecial   #37; //Method java/lang/Object."<init>":()V
   4:   return
  LineNumberTable: 
   line 917: 0

  Signature: length = 0x2
   00 21 

}
```

**Comment by benlings** [comment]
Would it be possible to rename hashCode(T) to hashCodeOf(T) or hashOf(T)? Having used the .net equivalent to this (IEqualityComparer<T>), it was confusing have an overloaded method with very different purposes.

**Comment by axel22** [comment]
Fine with me, the topic should be raised on scala-internals though to reach a consensus on the exact name.

**Pull request merged**